src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -196,83 +198,138 @@ addStop (Env env) ref acc = Env env
       env.directives
   }
 
+notAtTop :: Env -> Env
+notAtTop (Env e) = Env e { atTop = false }
+
+floatAppLets
+  :: forall i o f
+   . Foldable f
+  => (Env -> BackendSemantics -> Array o -> BackendSemantics)
+  -> (i -> BackendSemantics)
+  -> (i -> BackendSemantics -> o)
+  -> (f i -> Array o)
+  -> Boolean
+  -> Env
+  -> BackendSemantics
+  -> f i
+  -> BackendSemantics
+floatAppLets ef hf tf f atTop env hd tl = do
+  if atTop then ef env hd (f tl)
+  else evalAssocLet env hd cont
+  where
+  cont ex hd' = go [] ex $ List.fromFoldable tl
+    where
+    ief = flip ef hd'
+    go acc ee = case _ of
+      List.Nil -> ief ee acc
+      List.Cons head tail -> evalAssocLet ee (hf head) (\e v -> go (acc <> [ tf head v ]) e tail)
```

Comment:
This function concerns me with 8 arguments. In it's usages I see lots of generic things like identity `pure $ pure ..` etc. The array version uses `NeutPrimUndefined` which makes me raise my eyebrows. I'm afraid it might be over abstracted. I personally am unable to reason about what this function is doing at a glance, and I feel like it's style and usages are not in keeping with the rest of the code in this module. I'm sure it's relatively straightforward in theory, but it also may be that it's fine to duplicate some of the logic a bit to keep call sites and implementations more concrete. My personal opinion is that extreme DRY is not worthwhile if it inhibits local reasoning.